### DISCUSSION TITLE
Custom Free Energy Cost Function for Mixed State in NetKet

### DISCUSSION THREAD

**Discussion by rezaquant** [netket/netket#2074] [General]
Hi all,

I'm exploring whether NetKet can support optimization of a free energy functional for a mixed-state ansatz of the form: F=⟨H⟩−T⋅S(ρ)
I'm assuming that it's possible to efficiently estimate the entropy S(ρ) via sampling from the model.

Is there a way in NetKet provide a custom cost function involving both energy and entropy?

Thanks in advance!

**Comment by gcarleo** [comment]
in general it's not possible to estimate the quantum entropy efficiently unfortunately

**Reply by rezaquant** [reply]
Fair point. But is it possible in NetKet to define a custom cost function that includes both energy and entropy? Imagine I have a restricted ansatz that allows me to efficiently estimate the entropy S(ρ).

**Comment by rezaquant** [comment]
@PhilipVinc can you comment on the above question ^^:  Having a a custom cost function in netket?

**Reply by PhilipVinc** [reply]
NetKEt provides nothing to do this out of the box.
You can take your variational state and write a function that computes the loss and gradients you want, by building on top of the standard things (vs.samples, `nk.jax.jacobian`, `nk.jax.apply_chunked` and friends.)
